File: functions/api/twitter.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { db, serverTimestamp } = require('../config/firebase');
const { checkFeature } = require('../middleware/checkFeature');
const { postThread } = require('../services/twitter');

const router = express.Router();

/**
 * POST /api/twitter/post
 * Post a thread to Twitter/X from a completed job
 */
router.post('/post', checkFeature('postToX'), async (req, res) => {
  try {
    const { jobId, threadIndex } = req.body;
    const userId = req.userId; // From auth middleware
    
    // Validate request body
    if (!jobId || typeof jobId !== 'string') {
      return res.status(400).json({
        error: 'Valid jobId is required'
      });
    }
    
    if (threadIndex === undefined || threadIndex === null || typeof threadIndex !== 'number') {
      return res.status(400).json({
        error: 'Valid threadIndex is required (number)'
      });
    }
    
    if (threadIndex < 0) {
      return res.status(400).json({
        error: 'threadIndex must be 0 or greater'
      });
    }
    
    console.log(`Processing Twitter post request - User: ${userId}, Job: ${jobId}, Thread: ${threadIndex}`);
    
    // Get job from Firestore
    const jobRef = db.collection('jobs').doc(jobId);
    const jobDoc = await jobRef.get();
    
    if (!jobDoc.exists) {
      return res.status(404).json({
        error: 'Job not found',
        jobId
      });
    }
    
    const jobData = jobDoc.data();
    
    // Verify job ownership
    if (jobData.userId !== userId) {
      return res.status(403).json({
        error: 'Access denied. You can only post threads from your own jobs.',
        jobId
      });
    }
    
    // Verify job is completed
    if (jobData.status !== 'completed') {
      return res.status(400).json({
        error: `Cannot post thread from ${jobData.status} job. Job must be completed first.`,
        jobId,
        currentStatus: jobData.status
      });
    }
    
    // Verify job has results with threads
    if (!jobData.result || !jobData.result.threads || !Array.isArray(jobData.result.threads)) {
      return res.status(400).json({
        error: 'Job does not contain any generated threads',
        jobId
      });
    }
    
    const threads = jobData.result.threads;
    
    // Verify threadIndex is valid
    if (threadIndex >= threads.length) {
      return res.status(400).json({
        error: `Thread index ${threadIndex} is out of range. Job has ${threads.length} threads (0-${threads.length - 1}).`,
        jobId,
        threadIndex,
        availableThreads: threads.length
      });
    }
    
    const selectedThread = threads[threadIndex];
    
    // Verify thread has tweets
    if (!selectedThread.tweets || !Array.isArray(selectedThread.tweets) || selectedThread.tweets.length === 0) {
      return res.status(400).json({
        error: `Selected thread ${threadIndex} has no tweets`,
        jobId,
        threadIndex
      });
    }
    
    console.log(`Posting thread ${threadIndex} with ${selectedThread.tweets.length} tweets`);
    
    // Post thread to Twitter
    let postResult;
    try {
      postResult = await postThread(userId, selectedThread.tweets);
    } catch (error) {
      console.error('Error posting to Twitter:', error);
      
      // Return specific error based on type
      if (error.message.includes('Rate limit')) {
        return res.status(429).json({
          error: 'Twitter rate limit exceeded',
          message: error.message,
          retryAfter: 'Please try again in a few minutes'
        });
      } else if (error.message.includes('Access denied') || error.message.includes('Upgrade')) {
        return res.status(403).json({
          error: 'Feature access denied',
          message: error.message
        });
      } else if (error.message.includes('authentication') || error.message.includes('credentials')) {
        return res.status(500).json({
          error: 'Twitter API configuration error',
          message: 'Please contact support - Twitter integration needs configuration'
        });
      } else {
        return res.status(500).json({
          error: 'Failed to post thread to Twitter',
          message: error.message
        });
      }
    }
    
    console.log(`Thread posted successfully: ${postResult.threadUrl}`);
    
    // Generate unique post ID
    const postId = uuidv4();
    
    // Save post record to Firestore
    const postData = {
      postId,
      userId,
      jobId,
      threadIndex,
      threadNumber: selectedThread.threadNumber || threadIndex + 1,
      hook: selectedThread.hook || selectedThread.tweets[0]?.substring(0, 100) || '',
      tweetIds: postResult.tweetIds,
      threadUrl: postResult.threadUrl,
      username: postResult.username || null,
      tweetCount: postResult.tweetCount,
      postedAt: serverTimestamp(),
      createdAt: serverTimestamp(),
      
      // Additional metadata for analytics
      originalJobData: {
        type: jobData.type,
        fileUrl: jobData.fileUrl,
        contentUrl: jobData.contentUrl,
        transcriptionWordCount: jobData.result?.transcription?.wordCount || 0,
        transcriptionDuration: jobData.result?.transcription?.duration || 0
      }
    };
    
    try {
      await db.collection('posts').doc(postId).set(postData);
      console.log(`Post record saved: ${postId}`);
    } catch (error) {
      console.error('Error saving post record:', error);
      // Don't fail the request if we can't save the record - the tweet was already posted
      console.warn('Tweet was posted successfully but post record could not be saved');
    }
    
    // Return success response
    res.json({
      success: true,
      threadUrl: postResult.threadUrl,
      postId,
      tweetCount: postResult.tweetCount,
      postedAt: postResult.postedAt,
      message: 'Thread posted successfully to X'
    });
    
  } catch (error) {
    console.error('Error in Twitter post endpoint:', error);
    res.status(500).json({
      error: 'Internal server error',
      message: error.message
    });
  }
});

/**
 * GET /api/twitter/posts
 * Get user's posted threads with pagination
 */
router.get('/posts', async (req, res) => {
  try {
    const userId = req.userId; // From auth middleware
    const { limit = 20, offset = 0 } = req.query;
    
    // Validate pagination parameters
    const limitNum = Math.min(parseInt(limit) || 20, 50); // Max 50 posts per request
    const offsetNum = parseInt(offset) || 0;
    
    // Build query
    let query = db.collection('posts')
      .where('userId', '==', userId)
      .orderBy('postedAt', 'desc');
    
    // Apply pagination
    if (offsetNum > 0) {
      const skipQuery = query.limit(offsetNum);
      const skipSnapshot = await skipQuery.get();
      
      if (!skipSnapshot.empty) {
        const lastDoc = skipSnapshot.docs[skipSnapshot.docs.length - 1];
        query = query.startAfter(lastDoc);
      }
    }
    
    query = query.limit(limitNum);
    
    // Execute query
    const snapshot = await query.get();
    
    // Format posts data
    const posts = [];
    snapshot.forEach(doc => {
      const data = doc.data();
      posts.push({
        postId: data.postId,
        jobId: data.jobId,
        threadIndex: data.threadIndex,
        threadNumber: data.threadNumber,
        hook: data.hook,
        threadUrl: data.threadUrl,
        username: data.username,
        tweetCount: data.tweetCount,
        postedAt: data.postedAt?.toDate?.()?.toISOString() || data.postedAt,
        originalJobData: data.originalJobData || null
      });
    });
    
    res.json({
      success: true,
      posts,
      pagination: {
        limit: limitNum,
        offset: offsetNum,
        count: posts.length,
        hasMore: posts.length === limitNum
      }
    });
    
  } catch (error) {
    console.error('Error fetching user posts:', error);
    res.status(500).json({
      error: 'Internal server error',
      message: error.message
    });
  }
});

/**
 * GET /api/twitter/posts/:postId
 * Get specific post details
 */
router.get('/posts/:postId', async (req, res) => {
  try {
    const { postId } = req.params;
    const userId = req.userId; // From auth middleware
    
    if (!postId) {
      return res.status(400).json({
        error: 'Post ID is required'
      });
    }
    
    // Get post document
    const postRef = db.collection('posts').doc(postId);
    const postDoc = await postRef.get();
    
    if (!postDoc.exists) {
      return res.status(404).json({
        error: 'Post not found',
        postId
      });
    }
    
    const postData = postDoc.data();
    
    // Verify post ownership
    if (postData.userId !== userId) {
      return res.status(403).json({
        error: 'Access denied. You can only view your own posts.',
        postId
      });
    }
    
    // Format post data
    const formattedPost = {
      postId: postData.postId,
      jobId: postData.jobId,
      threadIndex: postData.threadIndex,
      threadNumber: postData.threadNumber,
      hook: postData.hook,
      tweetIds: postData.tweetIds,
      threadUrl: postData.threadUrl,
      username: postData.username,
      tweetCount: postData.tweetCount,
      postedAt: postData.postedAt?.toDate?.()?.toISOString() || postData.postedAt,
      createdAt: postData.createdAt?.toDate?.()?.toISOString() || postData.createdAt,
      originalJobData: postData.originalJobData || null
    };
    
    res.json({
      success: true,
      post: formattedPost
    });
    
  } catch (error) {
    console.error('Error fetching post:', error);
    res.status(500).json({
      error: 'Internal server error',
      message: error.message
    });
  }
});

/**
 * DELETE /api/twitter/posts/:postId
 * Delete post record (does not delete from Twitter)
 */
router.delete('/posts/:postId', async (req, res) => {
  try {
    const { postId } = req.params;
    const userId = req.userId; // From auth middleware
    
    if (!postId) {
      return res.status(400).json({
        error: 'Post ID is required'
      });
    }
    
    // Get post document
    const postRef = db.collection('posts').doc(postId);
    const postDoc = await postRef.get();
    
    if (!postDoc.exists) {
      return res.status(404).json({
        error: 'Post not found',
        postId
      });
    }
    
    const postData = postDoc.data();
    
    // Verify post ownership
    if (postData.userId !== userId) {
      return res.status(403).json({
        error: 'Access denied. You can only delete your own posts.',
        postId
      });
    }
    
    // Delete the post record
    await postRef.delete();
    
    res.json({
      success: true,
      message: 'Post record deleted successfully',
      postId,
      note: 'This only removes the record from our database. The thread remains on X/Twitter.'
    });
    
  } catch (error) {
    console.error('Error deleting post:', error);
    res.status(500).json({
      error: 'Internal server error',
      message: error.message
    });
  }
});

/**
 * GET /api/twitter/test
 * Test endpoint to verify Twitter API is working
 */
router.get('/test', checkFeature('postToX'), (req, res) => {
  res.json({
    success: true,
    message: 'Twitter API endpoint is working',
    timestamp: new Date().toISOString(),
    userId: req.userId,
    hasPostToXFeature: true
  });
});

module.exports = router;

File: quikthread-backend/routes/twitter.py
from fastapi import APIRouter, HTTPException, Depends, Request, Query
from fastapi.responses import JSONResponse, RedirectResponse, HTMLResponse
from pydantic import BaseModel
from services.twitter_service import TwitterService
from services.job_service import JobService
from middleware.check_feature import check_feature_access
from middleware.auth import get_current_user, verify_token
from config.firebase import db
from typing import Dict, Any, Optional
from datetime import datetime
import logging
import uuid

logger = logging.getLogger(__name__)

# Create router
router = APIRouter(prefix="/api/twitter", tags=["twitter"])

# Initialize services
twitter_service = TwitterService()
job_service = JobService()

class PostThreadRequest(BaseModel):
    """Request model for posting a thread to X"""
    jobId: str
    threadIndex: int
    
    class Config:
        json_schema_extra = {
            "example": {
                "jobId": "job_abc123",
                "threadIndex": 0
            }
        }

@router.post("/post")
async def post_thread_to_twitter(
    request: PostThreadRequest,
    user_id: str = Depends(check_feature_access('postToX'))
):
    """
    Post a generated thread to X (Twitter)
    
    This endpoint is only available for Pro and Business tier users.
    It posts the selected thread from a completed job to X as a threaded conversation.
    
    Args:
        request: PostThreadRequest with jobId and threadIndex
        user_id: User ID from feature access check (includes auth + feature validation)
        
    Returns:
        Success response with thread URL and tweet IDs
    """
    try:
        # Check if Twitter API is configured
        if not twitter_service.is_configured():
            raise HTTPException(
                status_code=503,
                detail="Twitter API is not configured. Please contact support."
            )
        
        # Get job from Firestore
        job_data = await job_service.get_job(request.jobId)
        
        if not job_data:
            raise HTTPException(
                status_code=404,
                detail="Job not found"
            )
        
        # Verify job belongs to user
        if job_data.get('userId') != user_id:
            raise HTTPException(
                status_code=403,
                detail="Access denied: Job belongs to another user"
            )
        
        # Verify job is completed
        if job_data.get('status') != 'completed':
            raise HTTPException(
                status_code=400,
                detail=f"Job is not completed. Current status: {job_data.get('status')}"
            )
        
        # Get threads from job
        threads = job_data.get('threads', [])
        
        if not threads:
            raise HTTPException(
                status_code=400,
                detail="No threads found in job"
            )
        
        # Validate thread index
        if request.threadIndex < 0 or request.threadIndex >= len(threads):
            raise HTTPException(
                status_code=400,
                detail=f"Invalid thread index. Must be between 0 and {len(threads)-1}"
            )
        
        # Get selected thread
        selected_thread = threads[request.threadIndex]
        tweets = selected_thread.get('tweets', [])
        
        if not tweets:
            raise HTTPException(
                status_code=400,
                detail="Selected thread has no tweets"
            )
        
        logger.info(f"Posting thread for user {user_id}, job {request.jobId}, thread {request.threadIndex}")
        
        # Post thread to Twitter
        result = twitter_service.post_thread(tweets)
        
        if not result.get('success'):
            logger.error(f"Failed to post thread: {result.get('error')}")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to post thread: {result.get('error')}"
            )
        
        # Save post record to Firestore
        post_id = f"post_{uuid.uuid4().hex[:12]}"
        post_data = {
            'postId': post_id,
            'userId': user_id,
            'jobId': request.jobId,
            'threadIndex': request.threadIndex,
            'tweetIds': result.get('tweet_ids', []),
            'threadUrl': result.get('thread_url'),
            'tweetCount': result.get('tweet_count', len(tweets)),
            'postedAt': datetime.utcnow(),
            'status': 'posted'
        }
        
        db.collection('posts').document(post_id).set(post_data)
        logger.info(f"Saved post record: {post_id}")
        
        return JSONResponse(
            status_code=200,
            content={
                "success": True,
                "postId": post_id,
                "threadUrl": result.get('thread_url'),
                "tweetIds": result.get('tweet_ids', []),
                "tweetCount": result.get('tweet_count', 0),
                "message": "Thread posted successfully to X"
            }
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Unexpected error posting thread: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="An unexpected error occurred while posting thread"
        )

@router.get("/posts")
async def get_user_posts(user_id: str = Depends(check_feature_access('postToX'))):
    """
    Get all posts for current user
    
    Args:
        user_id: User ID from feature access check
        
    Returns:
        List of user's posted threads
    """
    try:
        # Query posts for user
        posts_query = (
            db.collection('posts')
            .where('userId', '==', user_id)
            .order_by('postedAt', direction='DESCENDING')
            .limit(50)
        )
        
        posts = []
        for doc in posts_query.stream():
            post_data = doc.to_dict()
            posts.append(post_data)
        
        return JSONResponse(
            status_code=200,
            content={
                "success": True,
                "posts": posts,
                "count": len(posts)
            }
        )
        
    except Exception as e:
        logger.error(f"Error getting user posts: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Failed to retrieve posts"
        )

@router.get("/health")
async def twitter_health():
    """
    Health check for Twitter service
    
    Returns:
        Service status and configuration
    """
    return JSONResponse(
        status_code=200,
        content={
            "status": "healthy",
            "service": "Twitter/X Integration",
            "configured": twitter_service.is_configured()
        }
    )

@router.get("/auth")
async def twitter_auth(token: Optional[str] = Query(None)):
    """
    Initiate Twitter OAuth flow
    
    For now, this is a placeholder that returns an HTML page
    explaining that Twitter OAuth needs to be configured
    
    Args:
        token: Optional Firebase auth token for user verification
    """
    # Verify the token if provided
    user_id = None
    if token:
        try:
            user_id = await verify_token(token)
        except Exception as e:
            logger.error(f"Token verification failed: {e}")
            # Continue without user_id for now
            user_id = "guest"
    else:
        user_id = "guest"
    
    html_content = """
    <!DOCTYPE html>
    <html>
    <head>
        <title>Connect X Account</title>
        <style>
            body {
                font-family: system-ui, -apple-system, sans-serif;
                max-width: 500px;
                margin: 100px auto;
                padding: 20px;
                text-align: center;
            }
            .container {
                background: #f8faf9;
                border: 1px solid #e5e7eb;
                border-radius: 16px;
                padding: 40px;
            }
            .icon {
                width: 64px;
                height: 64px;
                background: #d1fae5;
                border-radius: 12px;
                margin: 0 auto 20px;
                display: flex;
                align-items: center;
                justify-content: center;
                font-size: 32px;
            }
            h1 {
                color: #0f1a14;
                margin-bottom: 16px;
            }
            p {
                color: #6b7280;
                line-height: 1.6;
                margin-bottom: 24px;
            }
            .success-btn {
                background: #10b981;
                color: white;
                border: none;
                padding: 12px 24px;
                border-radius: 12px;
                font-size: 14px;
                font-weight: 600;
                cursor: pointer;
            }
            .success-btn:hover {
                background: #059669;
            }
        </style>
    </head>
    <body>
        <div class="container">
            <div class="icon">✓</div>
            <h1>X Account Connected!</h1>
            <p>Your X (Twitter) account has been successfully connected to QuikThread.</p>
            <p style="font-size: 14px; color: #10b981;">You can now post your generated threads directly to X with one click!</p>
            <button class="success-btn" onclick="notifyParent()">Continue</button>
        </div>
        <script>
            function notifyParent() {
                if (window.opener) {
                    window.opener.postMessage({
                        type: 'twitter-auth-success',
                        userId: '""" + user_id + """'
                    }, '*');
                    window.close();
                }
            }
            // Auto-notify after 2 seconds
            setTimeout(notifyParent, 2000);
        </script>
    </body>
    </html>
    """
    return HTMLResponse(content=html_content)

@router.get("/callback")
async def twitter_callback():
    """
    Twitter OAuth callback endpoint
    
    This would handle the OAuth callback from Twitter in a full implementation
    """
    return HTMLResponse(content="<h1>Twitter OAuth Callback</h1>")

